### PR TITLE
Implement missing form template GPS-related insertion tags

### PR DESCRIPTION
**Pull request by n2ygk**
Fixes: #328 

Implements several missing insertion tags for form templates (some of which are not documented other than by reading the HTML):

- {GPS} -  degrees and decimal minutes with Northing and Easting: 46-22.77N 121-35.01W
- {GPS_DECIMAL} - decimal degrees with Northing and Easting: 46.3795N 121.5835W
- {GPS_SIGNED_DECIMAL} - signed decimal degrees: 41.1234 -73.4567
- {Latitude} and {latitude} - signed decimal latitude degrees: 41.1234
- {Longitude} and {longitude} - signed decimal longitude degrees: -73.4567
- {GridSquare} - maidenhead grid square locator: FN31cd
- {GPSValid} - "YES " or "NO " to indicate GPS position is valid

Some of these values are substituted in forms but not used (like GridSquare).

### CODE REVIEW

**Comment by n2ygk** [comment]
@xylo04 I thought I'd give it a try.

**Comment by n2ygk** [comment]
(ignore those extra commits -- I was having issues pushing from my virtualbox...)

File: internal/forms/forms.go
Comment (xylo04) [inline internal/forms/forms.go:724]: Within the Manager struct (most of the functions here are actually methods on that struct), the config object is available. So as long as that's in scope, getting this value should be as simple as m.config.GPSd.Addr.

File: internal/forms/forms.go
Comment (xylo04) [inline internal/forms/forms.go:779]: I would suggest separating the tasks of getting the position from the GPS device vs formatting the position into different methods. Since we're having to do some costly communication with the GPS device, doing that three times seems like overkill.

**Review by xylo04 — CHANGES_REQUESTED** [review]
@n2ygk this looks like a great first crack! Thanks for putting this together. Does it work for you locally? If so, that's super exciting!

File: internal/forms/forms.go
Comment (n2ygk) [reply internal/forms/forms.go:712]: Actually I tried this at first and there's a local version of Config that appears not to have GPSd.... 

https://github.com/la5nta/pat/blob/ccb493474debb2a84ee2ab60b292327426c86f8a/internal/forms/forms.go#L55-L62

**Comment by n2ygk** [comment]
@xylo04 This is working properly for the GPS form but not the Winlink Check In form. Not sure why just yet other than suspecting I'm either not formatting the signed decimal string correctly or the form is buggy. The LAT gets filled in properly but the LON doesn't.

See attached before and after screen shots for the GPS Position report form. In before, the tag has not been substituted and still showed as `{GPS}` while in the after, it has my position.
Before:
before

After:
after

Clicking on "Click to use connected GPS LAT LON" does the right thing and fills in all the other fields on the form.

**Comment by n2ygk** [comment]
I got the Winlink Check In form working by reverse-engineering and finding several undocumented insertion tags which I've added in [32e7106](https://github.com/la5nta/pat/pull/330/commits/32e71069c05954e2003caa5442e5038ef9fd8130).

There's a lot of random capitalization going on and some tags that are both supplied by the HTTP server and calculated in the form -- notably Grid Square.

File: internal/forms/forms.go
Comment (xylo04) [reply internal/forms/forms.go:712]: Ah, I see you're right. In that case you'll need to add a new field to forms.Config to hold the GPSd address, and plumb it though in main.go where we call forms.NewManager.

File: internal/forms/forms.go
Comment (xylo04) [inline internal/forms/forms.go:841]: Good find, but... yeesh.

**Review by xylo04 — CHANGES_REQUESTED** [review]
Nice improvements! Getting closer.

**Comment by martinhpedersen** [comment]
Thanks for working on this @n2ygk 🙂 

It looks like you haven't configured your editor to run `go fmt`? If you haven't, please remember `go fmt` before squashing and finalizing this PR. Thanks!

PS: We have a [CONTRIBUTING.md](https://github.com/la5nta/pat/blob/master/CONTRIBUTING.md) with some guidance to first time contributors. It's not the best, but worth the read hopefully 🙂

File: internal/forms/forms.go
Comment (n2ygk) [reply internal/forms/forms.go:741]: I googled "golang equivalent of enum"

File: internal/forms/forms.go
Comment (n2ygk) [reply internal/forms/forms.go:863]: Yeah, ugh.

**Comment by n2ygk** [comment]
> Thanks for working on this @n2ygk 🙂
> 
> It looks like you haven't configured your editor to run `go fmt`? If you haven't, please remember `go fmt` before squashing and finalizing this PR. Thanks!
> 
> PS: We have a [CONTRIBUTING.md](https://github.com/la5nta/pat/blob/master/CONTRIBUTING.md) with some guidance to first time contributors. It's not the best, but worth the read hopefully 🙂

@martinhpedersen: no sorry I didn't run `go fmt` but when I do the changes are to unrelated files:
```
vagrant@debian11:~/src/pat$ go fmt
http_multipart.go
riglist.go
```
Am I missing something? I can certainly rebase. Will do that now. Thanks for the reminder.

**Comment by martinhpedersen** [comment]
> ... but when I do the changes are to unrelated files:
> 
> ```
> vagrant@debian11:~/src/pat$ go fmt
> http_multipart.go
> riglist.go
> ```
> Am I missing something?

These seems to be related to a change in how build tags are handled (>Go 1.17). I will fix those 👍 

Running `go fmt` will only modify files in the current directory. You could use `go fmt ./...` to format sub-packages as well, but most Go developers configure their editor to run `go fmt` on the target file on save. Most Go-enabled IDEs will do this for you out of the box, and plain old editors like `vim` and `emacs` have various plugins to support this. (I'm a vim user myself, and use [vim-go](https://github.com/fatih/vim-go)).

**Comment by n2ygk** [comment]
@martinhpedersen Ah. Thanks for the pointer. My editor is Emacs and my go knowledge is zero.
I found a few others besides the one I touched. Looks like the same build tags issue.  I'll fix forms.go and push.
```
vagrant@debian11:~/src/pat$ go fmt ./...
http_multipart.go
riglist.go
internal/forms/forms.go
internal/osutil/rlimit_unix.go
internal/osutil/rlimit_freebsd.go
internal/osutil/rlimit_windows.go
```

**Comment by n2ygk** [comment]
@martinhpedersen fixed in my .emacs: `(add-hook 'before-save-hook #'gofmt-before-save)"`

**Comment by martinhpedersen** [comment]
> fixed in my .emacs: `(add-hook 'before-save-hook #'gofmt-before-save)"`

Nice! 🙂 👍 

One more thing! We generally merge PRs to `develop` to keep `master` on track with the latest release. If you don't mind, please update the base branch of this PR and rebase on top of `develop`. This is one of the things missing in CONTRIBUTORS.md (note to self). Sorry about that.

Thanks!

**Comment by martinhpedersen** [comment]
@n2ygk - Are you still working on this, or waiting for a new review and/or merge?

**Comment by n2ygk** [comment]
> @n2ygk - Are you still working on this, or waiting for a new review and/or merge?

I'm done. Waiting for review/merge.

File: internal/forms/forms.go
Comment (martinhpedersen) [inline internal/forms/forms.go:753]: It's preferable to define variables as local as possible. I think this is a symptom of the complexity introduced by accepting the style parameter instead of having multiple implementing functions for the different styles.

PS: When declaring lots of variables like this, it is a bit cleaner to use this syntax:

```go
var (
   northing string
   easting string
   latDegrees int
   latMinutes float64
   lonDegrees int
   lonMinutes float64
)
```

File: internal/forms/forms.go
Comment (martinhpedersen) [inline internal/forms/forms.go:795]: We're already depending on github.com/pd0mz/go-maidenhead for doing maidenhead calculation. Maybe it provides some of this functionality already?

File: internal/forms/forms.go
Comment (martinhpedersen) [inline internal/forms/forms.go:877]: TODO?

**Review by martinhpedersen — CHANGES_REQUESTED** [review]
Thanks for working on this @n2ygk 😄 

I think we're almost ready to merge this. Please see my inline comments.

Thank you!

File: internal/forms/forms.go
Comment (n2ygk) [reply internal/forms/forms.go:719]: yep

File: internal/forms/forms.go
Comment (n2ygk) [reply internal/forms/forms.go:730]: I kinda like seeing it in the console log... but sure, I can remove it.

Comment (n2ygk) [reply internal/forms/forms.go:730]: Also, it definitively ends the "Waiting for position from GPSd ..." which can take up to 3 seconds before either succeeding or failing.

File: internal/forms/forms.go
Comment (n2ygk) [reply internal/forms/forms.go:747]: I thought I was being so fancy google what the go version of a C enum was... ;-)

Comment (n2ygk) [reply internal/forms/forms.go:747]: Some of the paths reuse code, so...

File: internal/forms/forms.go
Comment (n2ygk) [reply internal/forms/forms.go:877]: I went through the forms and found a bunch of undocumented insertion tags. One if them was `InternetAvailable` which presumably somehow determines if Internet connectivity is currently in place. I didn't know how to determine that, and it wasn't really relevant to getting the GPSd stuff working but I wanted to document that there is this (and perhaps other) insertion tags that still need to be implemented. I will remove this line and create an Issue instead.

Comment (n2ygk) [reply internal/forms/forms.go:877]: #338

File: internal/forms/forms.go
Comment (n2ygk) [reply internal/forms/forms.go:795]: Well, yes, there is! Thanks for that.

File: internal/forms/forms.go
Comment (n2ygk) [reply internal/forms/forms.go:746]: @martinhpedersen I've left this one as is. Let me know if that's OK.

Comment (martinhpedersen) [reply internal/forms/forms.go:746]: > I thought I was being so fancy googling what the go version of a C enum was... ;-)

Oh yes, your `iota` usage was spot on! 👍 

Sure! Let's keep it this way. No problem 🙂

File: internal/forms/forms.go
Comment (martinhpedersen) [inline internal/forms/forms.go:722]: Very dangerous to defer conn.Close before checking that error. If there is an error, changes are `conn` is nil and you will get a nil pointer panic once the deferred Close call is executed 😅

File: internal/forms/forms.go
Comment (martinhpedersen) [inline internal/forms/forms.go:734]:
```go
return conn.NextPosTimeout(3 * time.Second)
```

That would be even easier to read 🙂

File: internal/forms/forms.go
Comment (martinhpedersen) [inline internal/forms/forms.go:797]: Nice! 🎉

**Review by martinhpedersen — CHANGES_REQUESTED** [review]
Almost there!

**Comment by n2ygk** [comment]
I supposed I should rebase this.... Standby.

**Review by martinhpedersen — COMMENTED** [review]
Since our default config provides a [default gpsd address](https://github.com/la5nta/pat/blob/ccb493474debb2a84ee2ab60b292327426c86f8a/cfg/config.go#L300) this will always throw an error for users without GPSd.

I think we need a more graceful handling of this condition than printing the standard tcp dial error each time those users fills in a form template.

Maybe we should simply omit the error if it's a network dial error?

**Comment by n2ygk** [comment]
> Since our default config provides a [default gpsd address](https://github.com/la5nta/pat/blob/ccb493474debb2a84ee2ab60b292327426c86f8a/cfg/config.go#L300) this will always throw an error for users without GPSd.
> 
> I think we need a more graceful handling of this condition than printing the standard tcp dial error each time those users fills in a form template.
> 
> Maybe we should simply omit the error if it's a network dial error?

How about changing the default config to not list a gpsd address? It's kind of wasteful to attempt those dials irrespective of whether errors are being logged. I wouldn't want to not log errors if there is in fact expected to be a gpsd running.

**Comment by martinhpedersen** [comment]
I think we should avoid comprimising a good default value in our config just to keep from logging errors for users of these forms that don't care about using GPSd. The default config is very convenient for the other GPSd related features, requiring no config on most systems.

How about adding a `enable_forms` config option similar to `enable_http`? That would be a good for security as well, so we don't automatically expose a user's position without their concent.

https://github.com/la5nta/pat/blob/587580f4bce65b742bb01052ee082d3388ce0be0/cfg/config.go#L227

**Comment by martinhpedersen** [comment]
@n2ygk - I've added the additional gpsd config option in 07b56654aea82754cae0f506b070b1e2787ab990. You can either cherry-pick that and glue it together in this PR, or let me know and I'll fix the rest 🙂 

Also if you could add yourself to the Contributors list if READM.md, that would be great. Thanks!

**Comment by n2ygk** [comment]
@martinhpedersen I need a little help here. I've made the change but when trying to compile and run there appears to have been some change to the web js code that throws this error: `Uncaught ReferenceError: onFormLaunching is not defined` I guess I'm looking for the equivalent of `make install` that installed not only the pat binary but the web files....

**Comment by martinhpedersen** [comment]
@n2ygk - I think you should rebase again, and solve that conflict in README.md.

I don't fully understand why you would get such an error. Probably not related to this PR anyway, but I will check it out once you've rebased 🙂 

Thanks!!

**Comment by n2ygk** [comment]
> @n2ygk - I think you should rebase again, and solve that conflict in README.md.
> 
> I don't fully understand why you would get such an error. Probably not related to this PR anyway, but I will check it out once you've rebased 🙂
> 
> Thanks!!

I suspect this may be related to 7850c5f036cab4892429352a9318aaa685722d94 which I just found now that I've rebased;-)

**Comment by n2ygk** [comment]
@martinhpedersen OK so this is looking like it does the right thing now. It's being very quiet about the error message per your prior recommendation so the user will not realize that the GPSd stuff needs to add `"allow_forms": true`. I suppose a wiki page update is warranted as well. Let me know if you'd like me to submit that.

File: internal/forms/forms.go
Comment (martinhpedersen) [inline internal/forms/forms.go:714]: Since the error returned from this function is never used for anything but to check for success, I think we should replace this with a "ok" boolean. Then you could omit the check for "noPos" as well 🙂 

All these error messages could be useful for debugging, so I suggest logging them instead using our debug logger [debug.Printf](https://pkg.go.dev/github.com/la5nta/pat/internal/debug#Printf).

```
func (m *Manager) gpsPos() (pos gpsd.Position, ok bool) {

}
```

I'd be happy to do these last minute tweaks after merging if you prefer. Just say the word 🙂

**Review by martinhpedersen — COMMENTED** [review]
Once last thing... (see inline comment).

**Comment by martinhpedersen** [comment]
@n2ygk 
> I suppose a wiki page update is warranted as well. Let me know if you'd like me to submit that.

I agree! If you don't mind, yes please 😄

File: internal/forms/forms.go
Comment (n2ygk) [reply internal/forms/forms.go:715]: The code currently propagates the return from NextPosTimeout() so this will become quite a bit uglier as would the NoPos checking since gpsFmt is called with a Position.

I've added a couple of debug Printf of the gpsPos error.

Feel free to make the tweaks you suggest if you disagree.

Comment (martinhpedersen) [reply internal/forms/forms.go:715]: Sure, I see what you mean 🙂 

As a general rule of thumb I think it's better to *either* log or return an error, not both. It helps with readability and makes the responsibilities clearer IMO.

I think we should merge this now, and I'll see what can be done with regards to this afterwards.

Thank you 😄

**Review by martinhpedersen — APPROVED** [review]

**Comment by martinhpedersen** [comment]
Thanks for working on this @n2ygk! Highly appreciated 🙂

**Comment by n2ygk** [comment]
> @n2ygk
> 
> > I suppose a wiki page update is warranted as well. Let me know if you'd like me to submit that.
> 
> I agree! If you don't mind, yes please 😄

@martinhpedersen done: https://github.com/la5nta/pat/wiki/The-web-GUI

**Comment by olewsaa** [comment]
I've put together a Python script that request GPS position from the Signal K server. It respond with nn.dddd nn.dddd, but I included a routine that convert the output to a grid reference.  It would be benefical if Signal K could be added as a source. File is [get.grid.py](https://github.com/olewsaa/Yacht-computer/blob/master/pat/get.grid.py). Not all RPis have a local GPS attached.

**Pull request merged**
